lib/public_key/src/pubkey_policy_tree.erl

```diff
@@ -0,0 +1,414 @@
+                end;
+           (#'PolicyQualifierInfo'{policyQualifierId = ?'id-qt-cps',
+                                   qualifier = Qualifier}) ->
+                {uri, public_key:der_decode('CPSuri', Qualifier)}
+        end,
+    Collect = fun(#{qualifier_set := QSet}) ->
+                      lists:map(FormatQualifier, QSet)
+              end,
+    case collect_subtree_qualifiers(Collect, SubTree, Policy) of
+        [] ->
+            collect_subtree_qualifiers(Collect, SubTree, ?anyPolicy);
+        QSet ->
+            QSet
+    end.
+
+%% --------------------------------------------------------------------
+-spec constraint_policy_node_set(policy_tree()) -> [public_key:policy_node()].
+
+%%
+%% From: PKITS.pdf:
+%% If the valid_policy_tree includes a leaf node with a valid_policy
+%% of anyPolicy, then the user-constrained-policy-set is
+%% any-policy. Otherwise, the user-constrained-policy-set consists of
+%% the set containing the valid_policy from each node in the
+%% valid_policy_tree in which the valid_policy is not anyPolicy and
+%% the valid_policy of that node's parent is anyPolicy.  The
+%% authorities-constrained-policy-set may be computed using the same
+%% procedure on the valid_policy_tree before its intersection with the
+%% user-initial-policy-set has calculated
+%%
+%% That is this function caclualates the valid policy node
+%% set after being constrained either by the user or only by the
+%% authorities (certificate extensions).
+%% --------------------------------------------------------------------
+constraint_policy_node_set({}) ->
+    [];
+constraint_policy_node_set(Tree) ->
+    case any_leaves(Tree) of
+        [] ->
+            constrain(Tree);
+        AnyLeaves ->
+            AnyLeaves
+    end.
+
+%%--------------------------------------------------------------------
+-spec empty() -> policy_tree().
+%%--------------------------------------------------------------------
+empty() ->
+    {}.
+
+%%--------------------------------------------------------------------
+-spec in_set(Policy::public_key:oid(), Set::[public_key:policy_node()]) -> boolean().
+%%
+%% Is there a node with <Policy> in <Set>
+%%--------------------------------------------------------------------
+in_set(_, []) ->
+    false;
+in_set(Policy, [#{valid_policy := Policy} |_]) ->
+    true;
+in_set(Policy, [_ | Rest]) ->
+    in_set(Policy, Rest).
+
+%%--------------------------------------------------------------------
+-spec is_empty(policy_tree()) -> boolean().
+%%--------------------------------------------------------------------
+is_empty({}) ->
+    true;
+is_empty(_) ->
+    false.
+
+%%--------------------------------------------------------------------
+-spec map_leaves(policy_tree(), LeafFun::function()) -> policy_tree().
+%%
+%% Update all leaves as determined by <LeafFun>
+%%--------------------------------------------------------------------
+map_leaves({Parent, [{_, _}|_] = SubTree}, LeafFun) ->
```

Comment:
what is this: `{PolicyData, [{PolicyData, []}]}` ?
- if it is proper tree of depth 2 then map values should process it and for example add a leave creating tree of depth 3
- if it is not a proper tree - probably return something like: `{error, leaf_not_found}`
- currently, it does none of the things - it returns unchanged input tree of depth 2 ...

I think it would simplify things if leaf node (or just single node tree) is represented with same tree node tuple.
So leaf node looks like:
`{PolicyData, []}`

And plain PolicyData is not recognized as tree node at all ... discussion wanted here.

Reply:
Plain PolicyData is not recognized as tree node only as the child part of a node.  In a node when the child/children are leaf(s) all children will not have any children of their own and it possible to make them always represented with an empty list, but I think it is unnecessary to create all that extra representation. This is not a general data-structure tree, it is a specific policy tree. The root node need to specify the empty children at start, it will never have no children if we have a valid tree, it is just the starting point. We do not need to worry about invalid trees as the trees are always created by our code that we should trust and not be defensive about.

Reply:
`{PolicyData, [{PolicyData, []}]}` - isn't it a tree node with no children aka tree `leaf` ?
- it will match with `{Parent, [{_, _}| _]}` which intent is to detect if we are on leaves level or not.

let's park this discussion for a while. i need to read more.

Reply:
When you reach the leaf children of a branch they will be a list of [policy_node()] not a list of tree nodes.